app/src/ui/create-tag/create-tag-dialog.tsx

```diff
@@ -0,0 +1,184 @@
+import * as React from 'react'
+
+import { Repository } from '../../models/repository'
+import { Dispatcher } from '../dispatcher'
+import { sanitizedRefName } from '../../lib/sanitize-ref-name'
+import { TextBox } from '../lib/text-box'
+import { Row } from '../lib/row'
+import { Dialog, DialogError, DialogContent, DialogFooter } from '../dialog'
+
+import { OkCancelButtonGroup } from '../dialog/ok-cancel-button-group'
+import { startTimer } from '../lib/timing'
+import { Octicon, OcticonSymbol } from '../octicons'
+import { Ref } from '../lib/ref'
+
+interface ICreateTagProps {
+  readonly repository: Repository
+  readonly dispatcher: Dispatcher
+  readonly onDismissed: () => void
+  readonly targetCommitSha: string
+  readonly initialName?: string
+}
+
+interface ICreateTagState {
+  readonly currentError: Error | null
+  readonly proposedName: string
+  readonly sanitizedName: string
+
+  /**
+   * Note: once tag creation has been initiated this value stays at true
+   * and will never revert to being false. If the tag creation operation
+   * fails this dialog will still be dismissed and an error dialog will be
+   * shown in its place.
+   */
+  readonly isCreatingTag: boolean
+
+  readonly localTags: Set<string>
+}
+
+const MAX_TAG_NAME_LENGTH = 245
+
+/** The Create Tag component. */
+export class CreateTag extends React.Component<
+  ICreateTagProps,
+  ICreateTagState
+> {
+  public constructor(props: ICreateTagProps) {
+    super(props)
+
+    this.state = {
+      currentError: null,
+      proposedName: props.initialName || '',
+      sanitizedName: '',
+      isCreatingTag: false,
+      localTags: new Set(),
+    }
+  }
+
+  public async componentDidMount() {
+    if (this.state.proposedName.length > 0) {
+      this.updateTagName(this.state.proposedName)
+    }
```

Comment:
I think we can do the sanitization in the constructor, and I agree it'll much better 😄 

I don't think there should be any performance concerns about doing that!

Reply:
In 7a8f6ae61a37416bc3afab307a42f265f65cc92b I've refactor a bit the component to not hold the `currentError` as state anymore and calculate it on render (it should be fine in terms of performance since this component only gets re-rendered whenever the tag name changes anyways).

With this I've been able to address both this suggestion and https://github.com/desktop/desktop/pull/9440#discussion_r409124573 😄